### PR TITLE
Implement fixed-length lists for BeaconState

### LINKED ISSUE #367: Fixed-length SSZ lists
## Description

During #364 I have come across an issue with how we deal with spec "constants".

## Summary

When SSZ decoding some bytes into a `BeaconState`, we must know the length of any fixed-length list (e.g., `latest_randao_mixes`, `current_crosslinks`, etc) in order to decode it.

Presently we have these lists as variable-length `Vec` and we have a chicken/egg problem where we don't know how long the list is until we decode it and we can't decode the list without knowing how long it is.

## Detail

The following `BeaconState` fields have this issue:

```python
    ...
    'latest_randao_mixes': ['bytes32', LATEST_RANDAO_MIXES_LENGTH],
    ...
    'current_crosslinks': [Crosslink, SHARD_COUNT],
    'previous_crosslinks': [Crosslink, SHARD_COUNT],
    'latest_block_roots': ['bytes32', SLOTS_PER_HISTORICAL_ROOT],
    'latest_state_roots': ['bytes32', SLOTS_PER_HISTORICAL_ROOT],
    'latest_active_index_roots': ['bytes32', LATEST_ACTIVE_INDEX_ROOTS_LENGTH],
    'latest_slashed_balances': ['uint64', LATEST_SLASHED_EXIT_LENGTH],  # Balances slashed at every withdrawal period
    ...
```

Here are some solutions to this issue:

1. Hard-code the required list-lengths as `const`.
2. Modify our SSZ impl such that some "context" is passed around allowing us to determine how big each list should be.
3. Implement a custom decode method for `BeaconState` such that some context is passed (e.g., `ChainSpec`) that allows determining the length of these fields (this is basically (2) but special-cased for `BeaconState`).
4. Make `BeaconState` generic across all list lengths and make them concrete when we instantiate a chain. There's a [rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dd30f998954155767c0d73772b681e2b) here to demonstrate.

## My Preference

I prefer (4) for the following reasons:
 - Maintains the ability to choose "constants" at runtime, as opposed to (1).
 - Keeps SSZ clean and doesn't require some nasty lookups to some "context" being passed around the decoder, as in (2).
 - Maintains a single, canonical method for de-serializing objects -- `Type::from_ssz_bytes(..)`. Doing (3) breaks this.

Downsides to (4):
 - Makes our `BeaconChain` definition a bit more complex (still peanuts compared to rust-libp2p, lol)
- Ergonomics -- all functions that deal with `BeaconState` need to be generic across those different array lengths. (As you can see in the playground link above, we can bundle them all into one trait, so you just need to specify `BeaconState<T> where T: Bundle`).

Additional upsides to (4):
 - We get a bit more type safety -- fixed-length lists can never be the wrong size (that's possible now).
 - Some more stuff lives on the stack instead of the heap (not a big deal.. huge lists might need to live in the heap anyway)

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

Resolves #367

## Proposed Changes

- Adds a `EthSpec` trait which allows for de-serializing fixed-length SSZ lists.
- Laboriously threads `EthSpec` through all existing code. 
- Commit 49c92ef removes the following items:
  - The `test_harness` crate.
  - The `block_proposer` and `attester` crates (only used by `test_harness`)
  - `network` tests that relied upon `test_harness`. 
- Commit 77c4b6e removes the YAML state-processing testing code, and 5988360  removes the build scripts.
- Commit 5988360 updates `.travis.yml` so it only builds on `--release` for nightly and beta (i.e., it does nothing else, it doesn't build on stable and it doesn't actually run tests). This is just to give a cheap heads-up about any upcoming changes which might break our build. Eventually this will move over to gitlab. 

## Ergonomics

This change introduces an interesting situation. We now have two "specification" items:

- `EthSpec` (the new one)
- `ChainSpec`

`ChainSpec` contains a bunch of "constant" variables that we use during various parts of the program and `EthSpec` contains a bunch of _actual_ constants.

Care needs to be taken to ensure that `ChainSpec` is compatible with `EthSpec`, and as such I've made all the `ChainSpec` constructors to be `pub(crate)` (i.e., private for external crates) and added a `EthSpec::spec() -> ChainSpec` fn.

```rust
let spec = ChainSpec::foundation();  // No longer possible.

let spec = FoundationEthSpec::spec();  // The new way.
```

TBH, I don't really like this `ChainSpec` vs `EthSpec` scenario; I think it's a bit janky. That being said, I'm still not sure I have an ideal alternative. One might be to abandon `ChainSpec` and just have `EthSpec` return "constants" on demand. I suspect this will be more difficult to maintain, but maybe it's worth it?

Anyway, I'm at a point where what's here is good enough and I can't think of a solution immediately. My thinking is to roll with this for now, do the v0.6.1 updates and then, with all that new knowledge, see if we come up with a great solution. Keen to hear thoughts.

## Reasoning for removing some components

TL;DR for the following section: We're about to update to a new spec with a brand new testing architecture so I don't want to waste time refactoring things that this might replace.

### Reasoning for removing test harness

`test_harness` was exclusively using the `attester` and `proposer` crates, after the `validator_client` decided to make its own implementations. These crates were designed to be used by both the `validator_client` _and_ the `test_harness` -- they can be re-implemented better (simpler, more ergonomic) if they are exclusive to `test_harness`.

I found it doesn't make sense to spend the time updating `test_harness` right before we do a spec update an implement a whole new testing framework. After these new 

### Reasoning for removing state-processing tests

It's a significant amount of work to parse these tests as we now have fixed-length lists -- we can no longer simply parse a big struct of test cases in a single `serde_yaml` line because we need to know the contents of the struct in order to know the type of the struct (chicken/egg). Here's a rough flow of what needs to happen in order to parse the tests:

- Split the test cases into individual cases.
- Parse only the `config` section.
- Match the `config` section against a `BeaconStateTypes` definition.
- Parse the whole test doc with the knowledge of the fixed-list lengths gained in the above step.

As such, I don't see the value in doing this singificant refactor when we're about to go to a new spec version with a whole new testing framework.

### CODE REVIEW

**Comment by paulhauner** [comment]
I'm going to drop some regex patterns in here that'll be useful for @michaelsproul when he's merging his branch. I'll keep updating this as I go (no need for action).

### `(\()(([^)]|\n)*)(?=BeaconState)` 

Finds `BeaconState` and matches the first `(` preceeding it, without passing through a `)`.

**Comment by paulhauner** [comment]
@AgeManning and @michaelsproul it would be good for you guys to just have a look at the major design decisions here and give any feedback.

I'm going to push on with SSZ for now, perhaps we can get this merged early next week.

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 1]
```diff
@@ -83,30 +83,31 @@ impl BlockProcessingOutcome {
     }
 }
 
-pub struct BeaconChain<T: ClientDB + Sized, U: SlotClock, F: ForkChoice> {
+pub struct BeaconChain<T: ClientDB + Sized, U: SlotClock, F: ForkChoice, B: EthSpec> {
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: `B` for "Beacon"?

File: eth2/types/src/lib.rs [hunk 2]
```diff
@@ -27,7 +27,7 @@ pub mod pending_attestation;
 pub mod proposer_slashing;
 pub mod slashable_attestation;
 pub mod transfer;
-pub mod tree_hash_vector;
+// pub mod tree_hash_vector;
```
Comment (michaelsproul) [inline eth2/types/src/lib.rs]: ?

File: eth2/types/src/lib.rs [hunk 3]
```diff
@@ -66,7 +66,7 @@ pub use crate::slashable_attestation::SlashableAttestation;
 pub use crate::slot_epoch::{Epoch, Slot};
 pub use crate::slot_height::SlotHeight;
 pub use crate::transfer::Transfer;
-pub use crate::tree_hash_vector::TreeHashVector;
+// pub use crate::tree_hash_vector::TreeHashVector;
```
Comment (michaelsproul) [inline eth2/types/src/lib.rs]: Delete?

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 1 above)
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Yeah, the type used to be call `BeaconStateTypes`. It doesn't really make sense now, would be better as `E`.

I'll see if I can find all of these and fix them.

File: eth2/types/src/lib.rs (hunk 2 above)
Comment (paulhauner) [reply eth2/types/src/lib.rs]: Yep, good call

File: eth2/types/src/lib.rs (hunk 3 above)
Comment (paulhauner) [reply eth2/types/src/lib.rs]: Agreed

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: I don't know if it's worth spending time on, `B` is fine really

**Comment by michaelsproul** [comment]
This looks good to me!

I need to think about it some more, but I wonder if the right solution to the disparity between `ChainSpec` and `EthSpec` is to remove `ChainSpec` entirely and make every spec constant a compile-time constant. In the future we'll need to deal with hard forks changing spec constants, in which case we might need a bundle of `EthSpec` traits, like `EthSpecV0`, `EthSpecV1`, etc, with the `BeaconState` parametrised over all of them (so it can alter its behaviour according to the current slot, etc).

Regarding the removal of the test infrastructure, I mostly agree that updating it just before a spec update seems like a waste of time, although I would like to resurrect it soon, as I have a few ideas for tests independent of the EF ones for which the `test_harness` would be useful.

**Comment by paulhauner** [comment]
> I need to think about it some more, but I wonder if the right solution to the disparity between `ChainSpec` and `EthSpec` is to remove `ChainSpec` entirely and make every spec constant a compile-time constant. 

I agree, I was also leaning this way.

Thanks for the review and input!

**Comment by paulhauner** [comment]
@michaelsproul I have addressed all those issues!

I ended up changing `<B>` to `<E>` via a few find/replace. It was bugging me, couldn't let it go haha.
